frontend/i_types.py

```diff
@@ -0,0 +1,298 @@
+from abc import ABCMeta, abstractmethod
+
+class Type(metaclass=ABCMeta):
+    """This class is the top-parent of all possible inferred types of a Python program.
+    Every type has its own class that inherits this class.
+    """
+
+    @abstractmethod
+    def is_subtype(self, t):
+        """Check if this type is a subtype of the parameter type.
+
+        Arguments:
+            type (Type): The type to check the subtype relationship against.
+        """
+        pass
+
+
+    @abstractmethod
+    def get_name(self):
+        """Get the name of the type"""
+        pass
+
+    # Two types are considered identical if their names exactly match.
+    def __eq__(self, other):
+        return self.get_name() == other.get_name()
+
+    def __hash__(self):
+        return hash(self.get_name())
+
+class TNone(Type):
+
+    def is_subtype(self, t):
+        return True
```

Comment:
Something like Maybe in Haskell. Basically, if there is a variable/field/... of type int, one could either allow or forbid putting None in that variable. If one does not allow that, one would need a different type Optional[int] which is allowed to contain either an integer or None. We should probably talk about what we want to do here.